tracing-test-utils/src/main/java/com/palantir/tracing/SpanRenderer.java

```diff
@@ -0,0 +1,382 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tracing;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectWriter;
+import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import com.google.common.graph.EndpointPair;
+import com.google.common.graph.GraphBuilder;
+import com.google.common.graph.MutableGraph;
+import com.google.common.hash.Hashing;
+import com.palantir.tracing.api.Span;
+import com.palantir.tracing.api.SpanObserver;
+import com.palantir.tracing.api.SpanType;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.Clock;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings({"PreferSafeLoggableExceptions", "Slf4jLogsafeArgs"}) // test-lib, no need for SafeArgs
+final class SpanRenderer implements SpanObserver {
+
+    private static final Logger log = LoggerFactory.getLogger(SpanRenderer.class);
+    private final Collection<Span> allSpans = new ArrayBlockingQueue<>(1000);
+
+    @Override
+    public void consume(Span span) {
+        allSpans.add(span);
+    }
+
+    @SuppressWarnings({"JavaTimeDefaultTimeZone", "BanSystemOut"}) // I actually want the system default time zone!
+    void output(String displayName, Path path) {
+        TimeBounds bounds = bounds(allSpans);
+
+        Map<String, List<Span>> spansByTraceId = allSpans.stream()
+                .collect(Collectors.groupingBy(Span::getTraceId));
+
+        Map<String, AnalyzedSpans> analyzedByTraceId = Maps.transformValues(spansByTraceId, SpanRenderer::analyze);
+
+        HtmlFormatter formatter = new HtmlFormatter(bounds);
+        StringBuilder sb = new StringBuilder();
```

Comment:
Can we operate on a Writer/PrintWriter instead of buffering the full output on heap twice?